File: packages/sdk/src/resources/sites.ts
import type { TheatricalHTTPClient } from '../http/client';
import { siteSchema, siteListResponseSchema, screenSchema } from '../types/site';
import type { Site, Screen } from '../types/site';

/**
 * Sites resource — cinema locations, screens, and geographic discovery.
 *
 * @example
 * ```typescript
 * const sites = await client.sites.list();
 * const active = sites.filter(s => s.isActive);
 * ```
 *
 * @example Geographic discovery
 * ```typescript
 * // Find cinemas within 10 km of central Wellington
 * const nearby = await client.sites.nearby(-41.2865, 174.7762, 10);
 * ```
 */
export class SitesResource {
  constructor(private readonly http: TheatricalHTTPClient) {}

  /**
   * List all cinema sites, with optional search query or geographic filter.
   * Response is validated at runtime using Zod.
   *
   * @param filters - Optional filters: `query` (text search), `latitude`/`longitude`/`radius`
   * @returns Array of cinema sites
   *
   * @example
   * ```typescript
   * const sites = await client.sites.list({ query: 'Embassy' });
   * ```
   */
  async list(filters?: {
    query?: string;
    latitude?: number;
    longitude?: number;
    radius?: number;
  }): Promise<Site[]> {
    const raw = await this.http.get<unknown>('/ocapi/v1/sites', {
      params: filters as Record<string, string | number | boolean | undefined>,
    });
    return siteListResponseSchema.parse(raw);
  }

  /**
   * Get a single cinema site by ID.
   * Response is validated at runtime using Zod.
   *
   * @param siteId - Vista site identifier
   * @returns The site record
   *
   * @example
   * ```typescript
   * const roxy = await client.sites.get('site_roxy_wellington');
   * console.log(roxy.screens.length); // number of auditoriums
   * ```
   */
  async get(siteId: string): Promise<Site> {
    const raw = await this.http.get<unknown>(`/ocapi/v1/sites/${siteId}`);
    return siteSchema.parse(raw);
  }

  /**
   * Get the screen (auditorium) configurations for a site.
   * Returns each screen's capacity, supported formats, and accessibility status.
   * Response is validated at runtime using Zod.
   *
   * @param siteId - Vista site identifier
   * @returns Array of screen configurations
   *
   * @example
   * ```typescript
   * const screens = await client.sites.screens('site_embassy_wellington');
   * const imax = screens.find(s => s.formats.includes('IMAX'));
   * ```
   */
  async screens(siteId: string): Promise<Screen[]> {
    const raw = await this.http.get<unknown>(`/ocapi/v1/sites/${siteId}/screens`);
    return screenSchema.array().parse(raw);
  }

  /**
   * Find cinema sites within a geographic radius.
   *
   * Returns sites sorted by distance from the given coordinates.
   * Uses Vista's OCAPI geographic search which accepts decimal lat/lng
   * and a radius in kilometres.
   *
   * @param latitude - Decimal latitude of the search centre (−90 to 90)
   * @param longitude - Decimal longitude of the search centre (−180 to 180)
   * @param radiusKm - Search radius in kilometres
   * @returns Sites within the radius, sorted nearest-first
   *
   * @example
   * ```typescript
   * // Find cinemas within 5 km of Roxy Cinema Wellington
   * const nearby = await client.sites.nearby(-41.3007, 174.7766, 5);
   * console.log(nearby.map(s => s.name));
   * ```
   *
   * @example Discover cinemas near central Auckland
   * ```typescript
   * const auckland = await client.sites.nearby(-36.8509, 174.7645, 10);
   * ```
   */
  async nearby(latitude: number, longitude: number, radiusKm: number): Promise<Site[]> {
    const raw = await this.http.get<unknown>('/ocapi/v1/sites', {
      params: { latitude, longitude, radius: radiusKm },
    });
    return siteListResponseSchema.parse(raw);
  }
}

File: packages/sdk-py/theatrical/resources/sites.py
"""Sites resource — cinema locations, screens, and geographic discovery."""

from __future__ import annotations

import builtins
from typing import TYPE_CHECKING, Optional

from pydantic import TypeAdapter

from theatrical.types.site import Screen, Site

if TYPE_CHECKING:
    from theatrical.http.client import TheatricalHttpProtocol

_site_list_adapter = TypeAdapter(builtins.list[Site])
_site_adapter = TypeAdapter(Site)
_screen_list_adapter = TypeAdapter(builtins.list[Screen])


class SitesResource:
    def __init__(self, http: TheatricalHttpProtocol) -> None:
        self._http = http

    async def list(
        self,
        *,
        query: Optional[str] = None,
        latitude: Optional[float] = None,
        longitude: Optional[float] = None,
        radius: Optional[float] = None,
    ) -> builtins.list[Site]:
        params: dict[str, str] = {}
        if query is not None:
            params["query"] = query
        if latitude is not None:
            params["latitude"] = str(latitude)
        if longitude is not None:
            params["longitude"] = str(longitude)
        if radius is not None:
            params["radius"] = str(radius)
        raw = await self._http.get("/ocapi/v1/sites", params=params or None)
        return _site_list_adapter.validate_python(raw)

    async def get(self, site_id: str) -> Site:
        raw = await self._http.get(f"/ocapi/v1/sites/{site_id}")
        return _site_adapter.validate_python(raw)

    async def screens(self, site_id: str) -> builtins.list[Screen]:
        raw = await self._http.get(f"/ocapi/v1/sites/{site_id}/screens")
        return _screen_list_adapter.validate_python(raw)

    async def nearby(
        self, latitude: float, longitude: float, radius_km: float
    ) -> builtins.list[Site]:
        raw = await self._http.get(
            "/ocapi/v1/sites",
            params={
                "latitude": str(latitude),
                "longitude": str(longitude),
                "radius": str(radius_km),
            },
        )
        return _site_list_adapter.validate_python(raw)
